storage/innobase/trx/trx0purge.cc

```diff
@@ -707,7 +707,16 @@ TRANSACTIONAL_TARGET void trx_purge_truncate_history()
           buf_pool.flush_hp.set(prev);
           mysql_mutex_unlock(&buf_pool.flush_list_mutex);
           if (!got_lock)
+          {
+            /* The functions buf_pool_t::release_freed_page() or
+            buf_do_flush_list_batch() may be right now holding
+            buf_pool.mutex and waiting to acquire
+            buf_pool.flush_list_mutex. Ensure that they can proceed,
+            to avoid extreme waits. */
+            mysql_mutex_lock(&buf_pool.mutex);
+            mysql_mutex_unlock(&buf_pool.mutex);
             bpage->lock.x_lock();
+          }
```

Comment:
This is an attempt to improve over 06280f233565fc030a406350a6f9b60e9c563d81 so that the undo tablespace truncation would not cause any prolonged blocking of the workload. In the first attempted run, we did not get any undo tablespace truncation triggered during the workload. During the shutdown, it happened, and took extremely long time (while ignoring `innodb_fast_shutdown=2`). I suspect that adding this wait could seriously increase the probability of `goto rescan`. That is, the linear scan of `buf_pool.flush_list` may degrade into a quadratic scan. A longer benchmark run is needed to determine if this is an improvement or something to be avoided.

Reply:
The newest update 98287bd2d6877969514469ebfbf27f9951b99657 is the same as 06280f233565fc030a406350a6f9b60e9c563d81, just rebased on the current 10.6 branch.